File: go_implement/iterative_merge_sort.go
package main

func MergeSortIterative(arr []int) []int {
	n := len(arr)
	temp := make([]int, n)

	for size := 1; size < n; size *= 2 {
		for left := 0; left < n-1; left += 2 * size {
			mid := min(left+size-1, n-1)
			right := min(left+2*size-1, n-1)

			mergeIter(arr, temp, left, mid, right)
		}
	}

	return arr
}

func mergeIter(arr, temp []int, left, mid, right int) {
	i, j, k := left, mid+1, left

	for i <= mid && j <= right {
		if arr[i] <= arr[j] {
			temp[k] = arr[i]
			i++
		} else {
			temp[k] = arr[j]
			j++
		}
		k++
	}

	for i <= mid {
		temp[k] = arr[i]
		i++
		k++
	}

	for j <= right {
		temp[k] = arr[j]
		j++
		k++
	}

	for i := left; i <= right; i++ {
		arr[i] = temp[i]
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: python_implement/iterative_merge_sort.py
"""Iterative implementation of merge sort algorithm."""


def merge_iter(arr, temp, left, mid, right):
    """Merges two sorted sub-arrays within the same array."""
    i = left
    j = mid + 1
    k = left

    while i <= mid and j <= right:
        if arr[i] <= arr[j]:
            temp[k] = arr[i]
            i += 1
        else:
            temp[k] = arr[j]
            j += 1
        k += 1

    while i <= mid:
        temp[k] = arr[i]
        i += 1
        k += 1

    while j <= right:
        temp[k] = arr[j]
        j += 1
        k += 1

    for i in range(left, right + 1):
        arr[i] = temp[i]


def merge_sort_iterative(arr):
    """Sorts an array using an iterative bottom-up merge sort algorithm."""
    '''We start by treating the array as a collections of cells size 1 and we merge them in sorted arrays'''
    n = len(arr)
    temp = [0] * n # temp is a table of size n we use to store merge results

    size = 1 # current size of a 'cell'
    while size < n:
        left = 0
        while left < n - 1:
            mid = min(left + size - 1, n - 1) # mid is the last index of the left sub-array
            right = min(left + 2 * size - 1, n - 1) # right is the last index of the right sub-array

            merge_iter(arr, temp, left, mid, right)

            left += 2 * size

        size *= 2

    return arr
